File: OrcScala/src/orc/lib/state/Semaphore.java
//
// Semaphore.java -- Java class Semaphore
// Project OrcScala
//
// $Id$
//
// Copyright (c) 2009 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.state;

import java.util.LinkedList;
import java.util.Queue;

import orc.TokenAPI;
import orc.error.runtime.TokenException;
import orc.values.sites.compatibility.Args;
import orc.values.sites.compatibility.DotSite;
import orc.values.sites.compatibility.EvalSite;
import orc.values.sites.compatibility.SiteAdaptor;

/**
 * @author quark
 */
@SuppressWarnings("hiding")
public class Semaphore extends EvalSite {

	@Override
	public Object evaluate(final Args args) throws TokenException {
		return new SemaphoreInstance(args.intArg(0));
	}

	protected class SemaphoreInstance extends DotSite {

		protected final Queue<TokenAPI> waiters = new LinkedList<TokenAPI>();
		protected final Queue<TokenAPI> snoopers = new LinkedList<TokenAPI>();
		protected int n;

		SemaphoreInstance(final int n) {
			this.n = n;
		}

		@Override
		protected void addMembers() {
			addMember("acquire", new SiteAdaptor() {
				@Override
				public void callSite(final Args args, final TokenAPI waiter) {
				  synchronized(SemaphoreInstance.this) {
					if (0 == n) {
						//FIXME:waiter.setQuiescent();
						waiters.offer(waiter);
						if (!snoopers.isEmpty()) {
							for (final TokenAPI snooper : snoopers) {
								//FIXME:snooper.unsetQuiescent();
								snooper.publish(signal());
							}
							snoopers.clear();
						}
					} else {
						--n;
						waiter.publish(signal());
					}
				  }
				}
			});
			addMember("acquirenb", new SiteAdaptor() {
				@Override
				public void callSite(final Args args, final TokenAPI waiter) {
                  synchronized(SemaphoreInstance.this) {
					if (0 == n) {
						waiter.halt();
					} else {
						--n;
						waiter.publish(signal());
					}
                  }
				}
			});
			addMember("release", new SiteAdaptor() {
				@Override
				public void callSite(final Args args, final TokenAPI sender) throws TokenException {
                  synchronized(SemaphoreInstance.this) {
					if (waiters.isEmpty()) {
						++n;
					} else {
						final TokenAPI waiter = waiters.poll();
						//FIXME:waiter.unsetQuiescent();
						waiter.publish(signal());
					}
					sender.publish(signal());
                  }
				}
			});
			addMember("snoop", new SiteAdaptor() {
				@Override
				public void callSite(final Args args, final TokenAPI snooper) throws TokenException {
                  synchronized(SemaphoreInstance.this) {
					if (waiters.isEmpty()) {
						//FIXME:snooper.setQuiescent();
						snoopers.offer(snooper);
					} else {
						snooper.publish(signal());
					}
                  }
				}
			});
			addMember("snoopnb", new SiteAdaptor() {
				@Override
				public void callSite(final Args args, final TokenAPI token) throws TokenException {
                  synchronized(SemaphoreInstance.this) {
					if (waiters.isEmpty()) {
						token.halt();
					} else {
						token.publish(signal());
					}
                  }
				}
			});
		}
	}
}

File: OrcScala/src/orc/lib/state/Semaphore.scala
//
// Semaphore.java -- Java class Semaphore
// Project OrcScala
//
// Copyright (c) 2018 The University of Texas at Austin. All rights reserved.
//
// Use and redistribution of this file is governed by the license terms in
// the LICENSE file found in the project's top-level directory and also found at
// URL: http://orc.csres.utexas.edu/license.shtml .
//

package orc.lib.state;

import scala.collection.mutable.Queue

import orc.{MaterializedCallContext, VirtualCallContext}
import orc.lib.state.types.SemaphoreType
import orc.run.distrib.{AbstractLocation, ClusterLocations, DOrcPlacementPolicy}
import orc.values.{FastObject, Signal}
import orc.values.sites.{
  FunctionalSite,
  NonBlockingSite,
  Site0Simple,
  TotalSite1Simple,
  TypedSite
}

/** @author quark
  * @author dkitchin
  */
object Semaphore
    extends TotalSite1Simple[Number]
    with TypedSite
    with FunctionalSite {
  override val inlinable = true
  def eval(arg: Number) = {
    val initialValue = arg.intValue
    if (initialValue >= 0) {
      new Semaphore.Instance(initialValue)
    } else {
      throw new IllegalArgumentException(
        "Semaphore requires a non-negative argument")
    }
  }

  def orcType() = {
    SemaphoreType.getBuilder
  }

  private val members =
    FastObject.members("acquire", "acquireD", "release", "snoop", "snoopD")

  class Instance(protected var n: Int)
      extends FastObject(members)
      with DOrcPlacementPolicy {
    instance =>

    require(n >= 0)

    val waiters = Queue[MaterializedCallContext]()
    var snoopers = Queue[MaterializedCallContext]()

    protected val values = Array(
      // "acquire" ->
      new Site0Simple {
        override val inlinable = true
        def eval(ctx: VirtualCallContext) = instance synchronized {
          if (0 == n) {
            val mat = ctx.materialize()
            mat.setQuiescent()
            waiters += mat
            if (snoopers.nonEmpty) {
              val oldSnoopers = snoopers.clone()
              snoopers.clear()
              oldSnoopers.foreach(_.publish(Signal))
            }
            ctx.empty
          } else {
            n -= 1
            ctx.publish(Signal)
          }
        }
      },
      // "acquireD" ->
      new Site0Simple with NonBlockingSite {
        override val inlinable = true
        def eval(ctx: VirtualCallContext) = instance synchronized {
          if (0 == n) {
            ctx.halt();
          } else {
            n -= 1
            ctx.publish(Signal);
          }
        }
      },
      // "release" ->
      new Site0Simple with NonBlockingSite {
        override val inlinable = true
        def eval(ctx: VirtualCallContext) = instance synchronized {
          val r = if (waiters.isEmpty) {
            n += 1
            ctx.empty
          } else {
            val waiter = waiters.dequeue()
            ctx.empty.publish(waiter, Signal)
          }
          r.publish(ctx, Signal)
        }
      },
      // "snoop" ->
      new Site0Simple {
        def eval(ctx: VirtualCallContext) = instance synchronized {
          if (waiters.isEmpty) {
            val mat = ctx.materialize()
            mat.setQuiescent()
            snoopers += mat
            ctx.empty
          } else {
            ctx.publish(Signal)
          }
        }
      },
      // "snoopD" ->
      new Site0Simple with NonBlockingSite {
        def eval(ctx: VirtualCallContext) = instance synchronized {
          if (waiters.isEmpty) {
            ctx.halt();
          } else {
            ctx.publish(Signal);
          }
        }
      }
    )

    def permittedLocations[L <: AbstractLocation](
        locations: ClusterLocations[L]): collection.immutable.Set[L] = {
      locations.hereSet
    }
  }
}
